### PR TITLE
Address warnings for possibly null array elements

### PR DESCRIPTION
**Pull request by cston**
As of https://github.com/dotnet/roslyn/pull/46405, the C# compiler considers constraint types for conversions involving type parameter types.

A warning is reported for the following for instance:
```C#
#nullable enable

interface I { }

class Program
{
    static void F0<T>(ref T x, ref T y)
    {
    }
    static void F<T>(ref T x, ref T y) where T : I
    {
        if (x == null) { }
        F0(ref x, ref y); // warning: x may be null
    }
}
```

This results in several new warnings building dotnet/runtime (see https://github.com/dotnet/roslyn/issues/46577).

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs [hunk 1]
```diff
@@ -473,8 +473,8 @@ private static int PickPivotAndPartition(Span<T> keys)
             {
                 if (pivot == null)
                 {
-                    while (Unsafe.IsAddressLessThan(ref leftRef, ref nextToLastRef) && (leftRef = ref Unsafe.Add(ref leftRef, 1)) == null) ;
-                    while (Unsafe.IsAddressGreaterThan(ref rightRef, ref zeroRef) && (rightRef = ref Unsafe.Add(ref rightRef, -1)) == null) ;
+                    while (Unsafe.IsAddressLessThan(ref leftRef!, ref nextToLastRef) && (leftRef = ref Unsafe.Add(ref leftRef, 1)) == null) ;
+                    while (Unsafe.IsAddressGreaterThan(ref rightRef!, ref zeroRef) && (rightRef = ref Unsafe.Add(ref rightRef, -1)) == null) ;
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: I'm not understanding why the compiler flags this. Why does the compiler think these methods could assign null to the ref?

Comment (cston) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: I believe the warning is for the value of `ref` argument when the method is called. On the second time through the loop, `leftRef` is assumed to be `null` based on `&& (leftRef = ...) == null`.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: I see.  Is the right fix here then to actually annotate both arguments with [AllowNull]?

Comment (cston) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: > Is the right fix here then to actually annotate both arguments with [AllowNull]?

The methods in `Unsafe` simply take an unconstrained `T` which seems like it should be sufficient for all callers. Should the `leftRef`, etc. locals be declared as `ref T?` here instead?

Comment (cston) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: > Should the `leftRef`, etc. locals be declared as `ref T?` here instead?

Perhaps not. That would just move the warnings to the initializers of those locals.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: > Should the leftRef, etc. locals be declared as ref T? here instead?

Wouldn't that be a problem if a `T t;` were passed in, with the compiler then complaining that the implementation might assign null to it?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs:477]: Right, hence my question about [AllowNull].  Does that make sense?

**Comment by stephentoub** [comment]
Thanks, @cston. Once this goes in, it should be ported to release/5.0 to unblock updating the compiler there as well. Thanks!
cc: @danmosemsft

**Comment by cston** [comment]
> I assume you validated adding the AllowNull does indeed also fix the issue?

Yes, adding `[AllowNull]` resolved the warnings.

**Review by safern — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs (hunk 1 above)
Comment (sharwell) [reply src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ArraySortHelper.cs]: This is similar to the `Volatile.Read` APIs, where ideally we could change `ref T` to `in T`, but that would be a breaking API change.

**Pull request merged**